### PR TITLE
Add support for generating debuginfo RPMs

### PR DESCRIPTION
**Pull request by kellyma2**
This change enables the creation and capture of debuginfo RPMs on
Fedora40 and CentOS7.

See:
https://docs.fedoraproject.org/en-US/packaging-guidelines/Debuginfo/

Fedora 40 expects the RPM contents to be located in a subdirectory
which is specified using the `buildsubdir` variable.  In order to
account for this, we need to tweak some of the location details if
debuginfo is enabled.

CentOS expects `buildsubdir` to have a value like `.` instead.
    
In both cases, we disable debugsource packages by ensuring that we
undefine `_debugsource_packages`, otherwise we'll try to generate them
alongside the debuginfo packages and will fail.

We only want this method of producing debuginfo to apply when we're
using the system `rpmbuild` because the underlying behaviour is
controlled by a combination of the rpmbuild version, macro
definitions, find-debuginfo.sh, and debugedit.  If we were to expand
this to use a hermetic debuginfo then a different approach might be
desirable.

### CODE REVIEW

**Review by cgrindel — COMMENTED** [review]
Looks reasonable to me. I just had a question about the rules_python version bump.

**Comment by aiuto** [comment]
I'm going to be slow to respond. Every hour of this week is booked.

File: MODULE.bazel [hunk 1]
```diff
@@ -7,7 +7,7 @@ module(
 
 # Do not update to newer versions until you need a specific new feature.
 bazel_dep(name = "rules_license", version = "0.0.4")
-bazel_dep(name = "rules_python", version = "0.24.0")
+bazel_dep(name = "rules_python", version = "0.31.0")
```
[the remark this answers is not in the document]
Comment (kellyma2) [reply MODULE.bazel:10]: It's broken on Fedora 40 - the underlying hermetic python in use is looking for libcrypt.so.1 which it can't find.  Both the test and example fail (and I expect other things would too).  The output is like:
```
INFO: Analyzed target //tests/rpm:test_golden_debuginfo_rpm_contents (58 packages loaded, 2730 targets configured).
ERROR: /src/tests/rpm/BUILD:562:8: MakeRpm tests/rpm/test_debuginfo_rpm-1-0..rpm failed: (Exit 127): make_rpm failed: error executing MakeRpm command (from target //tests/rpm:test_debuginfo_rpm) bazel-out/aarch64-opt-exec-ST-13d3ddad9198/bin/pkg/make_rpm '--name=test_debuginfo_rpm' '--rpmbuild=/usr/bin/rpmbuild' '--debuginfo_type=fedora40' ... (remaining 8 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
/home/test/.cache/bazel/_bazel_test/f8087e59fd95af1ae29e8fcb7ff1a3dc/sandbox/processwrapper-sandbox/8/execroot/_main/bazel-out/aarch64-opt-exec-ST-13d3ddad9198/bin/pkg/make_rpm.runfiles/rules_python~~python~python_3_11_aarch64-unknown-linux-gnu/bin/python3: error while loading shared libraries: libcrypt.so.1: cannot open shared object file: No such file or directory
Target //tests/rpm:test_golden_debuginfo_rpm_contents failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

**Comment by kellyma2** [comment]
> I'm going to be slow to respond. Every hour of this week is booked.

Thanks for the heads up

File: pkg/rpm_pfg.bzl [hunk 2]
```diff
@@ -1161,6 +1202,14 @@ pkg_rpm = rule(
                 [PackageSubRPMInfo],
             ],
         ),
+        "debuginfo": attr.bool(
+            doc = """Enable generation of debuginfo RPMs
+
+            For supported platforms this will enable the generation of debuginfo RPMs adjacent
+            to the regular RPMs.  Currently this is supported by Fedora 40 and CentOS7
```
Comment (aiuto) [inline pkg/rpm_pfg.bzl:1209]: . at the end of the docstring.

**Review by aiuto — COMMENTED** [review]
I have an overall question.
I was curious why you passed debuginfo_type via the toolchain, instead of just getting the os type in make_rpm.py.  Then I saw you need to change the stanza to add a .. in a path. sigh.

But it leads to the question of the name of `debuginfo_type`.  Is it really distinguishing the format of the debuginfo, or is it really a distinguisher of the target os?
If you could build for fedora from centos, should it be an attribute of the rule rather than picked up as a side effect in the toolchain?   

Are there other constraints or information about cross building which should make sense in the docstring? 

overall the code looks reasonable.

**Comment by kellyma2** [comment]
> But it leads to the question of the name of `debuginfo_type`. Is it really distinguishing the format of the debuginfo, or is it really a distinguisher of the target os? If you could build for fedora from centos, should it be an attribute of the rule rather than picked up as a side effect in the toolchain?
> 
> Are there other constraints or information about cross building which should make sense in the docstring?

tl;dr I don't think I can assert anything about cross building with this at all, unfortunately :( 

We're relying on the system rpmbuild for this which means that we're also relying on the various rpm macros and and supporting tools in place that belong to the system that we're building on.  `debuginfo_type` is a bit of a misnomer here, perhaps, but what it's referring to is the OS configuration for the aforementioned version of rpmbuild (incl macros, etc) *not* the target OS.  Initially I had allowed this to be selected in the module extension, but decided I just wanted to make this auto-detect so that we could at least hypothetically have the defined rules be system independant.

Can definitely clarify some of how this works in the docstring, however.

File: MODULE.bazel (hunk 1 above)
[the remark this answers is not in the document]
Comment (aiuto) [reply MODULE.bazel:10]: OK. So we'll do a major version and bzlmod bump when we release the next version.

**Review by aiuto — APPROVED** [review]
Can you just fix that period in the docstring and I'll merge.

**Comment by kellyma2** [comment]
> Can you just fix that period in the docstring and I'll merge.

Done

**Comment by aiuto** [comment]
Sorry for the  late merge. Hugely busy weekend.
